hoodie-common/src/test/java/com/uber/hoodie/common/table/log/HoodieLogFormatTest.java

```diff
@@ -1012,9 +1104,280 @@ public void testAvroLogRecordReaderWithMixedInsertsCorruptsAndRollback()
         .collect(Collectors.toList());
 
     HoodieCompactedLogRecordScanner scanner = new HoodieCompactedLogRecordScanner(fs, basePath,
-        allLogFiles, schema, "101", 10240L);
+        allLogFiles, schema, "101", 10240L, readBlocksLazily);
     assertEquals("We would read 0 records", 0,
         scanner.getTotalLogRecords());
   }
 
+  @Test
+  public void testMagicAndLogVersionsBackwardsCompatibility()
+      throws IOException, InterruptedException, URISyntaxException {
+    // Create the log file
+    Writer writer = HoodieLogFormat.newWriterBuilder().onParentPath(partitionPath)
+        .withFileExtension(HoodieLogFile.DELTA_EXTENSION).withFileId("test-fileid1")
+        .overBaseCommit("100").withFs(fs).build();
+    Schema schema = HoodieAvroUtils.addMetadataFields(getSimpleSchema());
+    List<IndexedRecord> records = SchemaTestUtil.generateHoodieTestRecords(0, 100);
+    Map<HoodieLogBlock.HeaderMetadataType, String> header = Maps.newHashMap();
+    header.put(HoodieLogBlock.HeaderMetadataType.INSTANT_TIME, "100");
+    writer.close();
+
+    // Write 1 with MAGIC and no log format version
+    // Append a log block to end of the log (mimics a log block with old format
+    // fs = FSUtils.getFs(fs.getUri().toString(), fs.getConf());
+    FSDataOutputStream outputStream = fs.append(writer.getLogFile().getPath());
+    // create a block with
+    outputStream.write(HoodieLogFormat.MAGIC);
+    outputStream.writeInt(HoodieLogBlockType.AVRO_DATA_BLOCK.ordinal());
+    // Write out a length that does not confirm with the content
+    records = SchemaTestUtil.generateHoodieTestRecords(0, 100);
+    header.put(HoodieLogBlock.HeaderMetadataType.SCHEMA, new String(HoodieAvroUtils.compress(schema.toString())));
+    HoodieAvroDataBlock dataBlock = new HoodieAvroDataBlock(records, schema);
+    byte [] content = dataBlock.getBytes(schema);
+    outputStream.writeInt(content.length);
+    // Write out some content
+    outputStream.write(content);
+    outputStream.flush();
+    outputStream.hflush();
+    outputStream.close();
+
+    writer = HoodieLogFormat.newWriterBuilder().onParentPath(partitionPath)
+        .withFileExtension(HoodieLogFile.DELTA_EXTENSION).withFileId("test-fileid1")
+        .overBaseCommit("100").withFs(fs).build();
+
+    // Write 2 with MAGIC_V2 and latest log format version
+    records = SchemaTestUtil.generateHoodieTestRecords(0, 100);
```

Comment:
please consider introducing some helpers to reduce redundant code here.. I am sure with sometime, these can be designed more concisely